core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregatorUtils.java

```diff
@@ -0,0 +1,89 @@
+import io.cdap.cdap.etl.api.relational.ExpressionFactory;
+import io.cdap.cdap.etl.api.relational.Relation;
+import io.cdap.cdap.etl.api.relational.RelationalTranformContext;
+import io.cdap.cdap.etl.api.relational.StringExpressionFactoryType;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Utility class for DedupAggregator.
+ */
+public class DedupAggregatorUtils {
+
+    public static DeduplicateAggregationDefinition generateAggregationDefinition
+            (RelationalTranformContext relationalTranformContext,
+             Relation relation,
+             DedupConfig.DedupFunctionInfo filterFunction,
+             List<String> uniqueFields) {
+        Optional<ExpressionFactory<String>> expressionFactory = relationalTranformContext.getEngine().
+                getExpressionFactory(StringExpressionFactoryType.SQL);
+        DeduplicateAggregationDefinition.FilterFunction aggFilterFunction = null;
+        if (filterFunction.getFunction() == DedupConfig.Function.MAX) {
+            aggFilterFunction = DeduplicateAggregationDefinition.FilterFunction.MAX;
+        } else if (filterFunction.getFunction() == DedupConfig.Function.MIN) {
+            aggFilterFunction = DeduplicateAggregationDefinition.FilterFunction.MIN;
+        } else {
+            return null;
+        }
+
+        Map<String, Expression> selectExpressions = new HashMap<>();
+        List<Expression> dedupExpressions = new ArrayList<>();
+        ExpressionFactory<String> stringExpressionFactory = expressionFactory.get();
+
+        Expression filterExpression = stringExpressionFactory.compile(getColumnName(relation,
+                filterFunction.getField(), stringExpressionFactory));
+
+        for (Schema.Field field : relationalTranformContext.getOutputSchema().getFields()) {
+            String columnName = getColumnName(relation, field.getName(), stringExpressionFactory);
+            selectExpressions.put(field.getName(), stringExpressionFactory.compile(columnName));
+        }
+
+
+        for (String uniqueField : uniqueFields) {
+            String columnName = getColumnName(relation, uniqueField, stringExpressionFactory);
+            dedupExpressions.add(stringExpressionFactory.compile(columnName));
+        }
+
+        return DeduplicateAggregationDefinition.builder()
+                .select(selectExpressions)
+                .dedupOn(dedupExpressions)
+                .filterDuplicatesBy(filterExpression, aggFilterFunction)
+                .build();
+    }
+
+    static String getColumnName(Relation relation,
+                                String name, ExpressionFactory<String> stringExpressionFactory) {
+        if (stringExpressionFactory.getCapabilities()
+                .contains(CoreExpressionCapabilities.CAN_GET_QUALIFIED_COLUMN_NAME)) {
+            return stringExpressionFactory.getQualifiedColumnName(relation, name).extract();
```

Comment:
How about making this return an `Expression`? Currently you will do `extract` and then `compile` most of the times. It's actually not needed.  You can still do `compile` if `CAN_GET_QUALIFIED_COLUMN_NAME` is not set

Reply:
okay so we can remove the check and have just this:

```
static ExtractableExpression<String> getColumnName(Relation relation,
                                                       String name, ExpressionFactory<String> stringExpressionFactory) {
            return stringExpressionFactory.getQualifiedColumnName(relation, name);
}
```

Reply:
What I mean is 
```
 if (stringExpressionFactory.getCapabilities()
                .contains(CoreExpressionCapabilities.CAN_GET_QUALIFIED_COLUMN_NAME)) {
            return stringExpressionFactory.getQualifiedColumnName(relation, name).extract();
} else {
            return stringExpressionFactory.compile(name);
}
```

Reply:
so I don't think I understand why we need to use `getQualifiedColumnName` if we can just `compile` and that's valid too.

Reply:
Because `getQualifiedColumnName` allows us to generate a column name that we are 100% certain is "safe" to use with the SQL engine.

In BigQuery's case, identifiers can be wrapped in backticks (`) as a way to ensure that the column name is safe to use.

So, for example:

```
select `thisísaçolumnName` from ...
```

Would be valid SQL in BQ because the column name is qualified.

Reply:
Yep. For CDAP more important cases where column name is a reserved word. We don't allow tons of things, bu t we do allow column names like `select` or `from`. Those fail without quoting, see https://cdap.atlassian.net/browse/PLUGIN-1003

Reply:
so in the else case (when we don't use `getQualifiedColumnName` and simply `compile`) shouldn't we throw an exception (since we are not 100% safe here)?

Reply:
It's not necessary to go that far.

Reply:
I see. Revised this in the new rev